pallets/interest-accrual/src/lib.rs

```diff
@@ -244,25 +281,77 @@ pub mod pallet {
 			accumulated_rate.checked_mul_int(normalized_debt)
 		}
 
-		fn calculate_accumulated_rate<Rate: FixedPointNumber>(
+		pub fn calculate_accumulated_rate<Rate: FixedPointNumber>(
 			interest_rate_per_sec: Rate,
 			accumulated_rate: Rate,
 			last_updated: Moment,
+			now: Moment,
 		) -> Result<Rate, DispatchError> {
 			// accumulated_rate * interest_rate_per_sec ^ (now - last_updated)
-			let time_difference_secs = Self::now()
+			let time_difference_secs = now
 				.checked_sub(last_updated)
 				.ok_or(ArithmeticError::Underflow)?;
 
-			checked_pow(interest_rate_per_sec, time_difference_secs as usize)
-				.ok_or(ArithmeticError::Overflow)?
-				.checked_mul(&accumulated_rate)
-				.ok_or(ArithmeticError::Overflow.into())
+			Ok(
+				checked_pow(interest_rate_per_sec, time_difference_secs as usize)
+					.and_then(|new_rate| new_rate.checked_mul(&accumulated_rate))
+					.ok_or(ArithmeticError::Overflow)?,
+			)
 		}
 
-		fn now() -> Moment {
+		pub fn now() -> Moment {
 			T::Time::now().as_secs()
 		}
+
+		pub fn reference_interest_rate(interest_rate_per_sec: T::InterestRate) {
+			Rate::<T>::mutate(interest_rate_per_sec, |rate| {
+				if let Some(rate) = rate {
+					rate.reference_count += 1;
+				} else {
+					*rate = Some(RateDetailsOf::<T> {
+						accumulated_rate: One::one(),
+						reference_count: 1,
+					});
+				}
+			})
+		}
+
+		pub fn unreference_interest_rate(interest_rate_per_sec: T::InterestRate) -> DispatchResult {
+			Rate::<T>::try_mutate(interest_rate_per_sec, |maybe_rate| {
+				if let Some(rate) = maybe_rate {
+					rate.reference_count -= 1;
+					if rate.reference_count == 0 {
+						*maybe_rate = None;
+					}
+					Ok(())
+				} else {
+					Err(Error::<T>::NoSuchRate.into())
+				}
+			})
+		}
+
+		pub fn upgrade_to_v1() -> Weight {
+			let mut weight = T::DbWeight::get().reads_writes(1, 1);
+			let version = Pallet::<T>::storage_version();
+			if version < 1 {
+				weight += migration::v1::migrate::<T>();
+			}
+			StorageVersion::<T>::set(1);
+			weight
+		}
+
+		pub fn remove_unused_rates() -> Weight {
```

Comment:
When is this used? 

I am not a big fan of translate as it is super slow.

Reply:
It's used as part of the migration. Since we only expect the migration to be run on Altair against a very small pool, I'm not too worried.